[Tachyon-1195] Implement persist command for TfsShell

manugoyal opened the pull request.

calvinjia commented:
@manugoyal This is only the metadata operation right?

manugoyal commented:
Oh whoops, didn't realize that the FileSystemMaster method wouldn't actually do the persisting.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/6457/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/6470/
Test PASSed.

apc999 commented:
@manugoyal , this only journals those operations. This JIRA aims to have an actually native way on the client side to read the tachyon data and simply dump it into under storage.

manugoyal commented:
The second commit adds the actual functionality that reads the file and writes it to the underfs. Is there something more I need to add?

apc999 commented on clients/unshaded/src/main/java/tachyon/client/FileSystemMasterClient.java:
```diff
@@ -467,4 +467,29 @@ public synchronized boolean unmount(TachyonURI tachyonPath) throws IOException {
+   * Persists the given Tachyon file with the given length
+   *
+   * @param fileId file id to persist
+   * @param length the length of the file
```
NIT: "the length of the file " in bytes?

apc999 commented on clients/unshaded/src/main/java/tachyon/client/FileSystemMasterClient.java:
```diff
@@ -467,4 +467,29 @@ public synchronized boolean unmount(TachyonURI tachyonPath) throws IOException {
   }
+
+  /**
+   * Persists the given Tachyon file with the given length
```
I thought our new style checker will raise error if this statement doesn't end with "."
surprise to see it passes the check.

apc999 commented on servers/src/main/java/tachyon/master/file/FileSystemMaster.java:
```diff
@@ -295,6 +295,14 @@ boolean persistFileInternal(long fileId, long length, long opTimeMs)
+    // Recursively traverse up the tree and mark all parent directories as persisted as well
+    long parentId = file.getParentId();
+    while (parentId != InodeTree.NO_PARENT) {
+      Inode parent = mInodeTree.getInodeById(parentId);
```
once catching a `parent` that is already persisted, do we still need to traverse up the tree?

apc999 replied:
@jsimsa , is this logic implemented also somewhere else? I will be surprised if we do have this logic before.

gpang replied:
Yeah, this should not be re-implemented here. You should use propagatePersisted()

apc999 commented on shell/src/main/java/tachyon/shell/TfsShell.java:
```diff
@@ -174,6 +178,61 @@ public void load(TachyonURI filePath) throws IOException {
+   * Persist a file or directory currently stored only in Tachyon to the UnderFileSystem
+   *
+   * @param filePath the TachyonURI path to persist to the UnderFileSystem
+   * @throws IOException
```
could you state when to throw this `IOException`?

apc999 commented on shell/src/main/java/tachyon/shell/TfsShell.java:
```diff
@@ -174,6 +178,61 @@ public void load(TachyonURI filePath) throws IOException {
+        for (FileInfo file : files) {
+          TachyonURI newPath = new TachyonURI(file.getPath());
+          try {
+            persist(newPath);
```
It looks like we call `persist` recursively, which is not entirely true as here `newPath` should be a file. So why don't we have a separate helper function `persistInternal` which only persists a file? In that way, the code looks less complicated, and it might be also easier to unit test.

manugoyal replied:
Can't `newPath` be a directory if we have multiple levels of nesting? It looks like `listStatus` only returns direct children of the path, right?

apc999 commented on shell/src/main/java/tachyon/shell/TfsShell.java:
```diff
@@ -174,6 +178,61 @@ public void load(TachyonURI filePath) throws IOException {
+          if (!ufs.exists(parentPath) && !ufs.mkdirs(parentPath, true)) {
+            throw new IOException("Failed to create " + parentPath);
+          }
+          OutputStream out = ufs.create(dstPath.getPath());
```
do we also need to register `out` to `closer`? Or if there is some internal mechanism happening inside `copy` so we don't have to close `out` explicitly, document it.

manugoyal replied:
Whoops yeah I have to register it.

apc999 commented:
thanks @manugoyal for your patch. Sorry that I missed your second commit. I reviewed your patch and left some comments.

yupeng9 commented on shell/src/main/java/tachyon/shell/TfsShell.java:
```diff
@@ -174,6 +178,61 @@ public void load(TachyonURI filePath) throws IOException {
+      TachyonFile fd = mTfs.open(filePath);
+      FileInfo fInfo = mTfs.getInfo(fd);
+      if (fInfo.isFolder) {
+        List<FileInfo> files = mTfs.listStatus(fd);
```
Shouldn't the support of persisting folder be pushed down in `TachyonFileSystem`? I can imagine such functionality can be useful in other places besides shell.

yupeng9 commented on shell/src/main/java/tachyon/shell/TfsShell.java:
```diff
@@ -174,6 +178,61 @@ public void load(TachyonURI filePath) throws IOException {
+          TachyonURI dstPath = new TachyonURI(fInfo.getUfsPath());
+          UnderFileSystem ufs = UnderFileSystem.get(dstPath.getPath(), mTachyonConf);
+          String parentPath = dstPath.getParent().getPath();
+          if (!ufs.exists(parentPath) && !ufs.mkdirs(parentPath, true)) {
```
I think the folder creation should be handled by `TachyonFileSystem`.

manugoyal replied:
Since it's a ufs operation, I think we have to do it manually. Seems like the same thing is done in `FileSystemMaster`

yupeng9 commented:
@manugoyal , thanks for adding this shell command. My high-level comment is try to push the implementation from shell into Tachyon FS client and Tachyon Master for reuse, so that the shell command can then delegate the call.

manugoyal commented:
@yupeng9 yeah @calvinjia and I were talking about that, and he felt that it might not be worth putting this functionality in the core client code until there's a significant feature need for it. Right now, this is mainly useful for people who accidentally forget to persist some files and want to persist a few in a one-off manner. Do you think there are valid use cases for putting this in the client code?

yupeng9 commented:
@manugoyal, I see. I am thinking from the usability perspective that there might be application developers want to call such method. But I also agree it's not necessary to be in the core client code. What about putting the logic of persisting folders in `TachyonFileSystemUtils`?

gpang commented on shell/src/main/java/tachyon/shell/TfsShell.java:
```diff
@@ -174,6 +178,61 @@ public void load(TachyonURI filePath) throws IOException {
+            throw new IOException("Failed to create " + parentPath);
+          }
+          OutputStream out = ufs.create(dstPath.getPath());
+          IOUtils.copy(in, out);
```
use copyLarge()?

gpang commented on shell/src/main/java/tachyon/shell/TfsShell.java:
```diff
@@ -174,6 +178,61 @@ public void load(TachyonURI filePath) throws IOException {
+        }
+        // Tell the master to persist the file
+        mTfs.persistFile(fd);
+        System.out.println(filePath + " persisted");
```
It would be nice to also print out the size of the data persisted, taken from the return value of copyLarge().

gpang commented:
Thanks for the improvement! I left a few comments. Also, please resolve conflicts.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/6570/
Test PASSed.

yupeng9 commented:
I left some minor comments. Otherwise LGTM.

gpang commented on clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemUtils.java:
```diff
@@ -134,4 +141,40 @@ public static boolean waitCompleted(final TachyonFileSystemCore tfs, final Tachy
+   * @throws FileDoesNotExistException if the given file does not exist
+   * @throws TachyonException if an unexpected Tachyon error occurs
+   */
+  public static long persistFile(TachyonFile file, TachyonURI dstPath, TachyonFileSystem tfs,
```
Is it possible to stay consistent with the other methods here, so that <code>tfs</code> is first, and it is of type <code>TachyonFileSystemCore</code>?

Also, is <code>dstPath</code> supposed to be related to <code>file</code>? Or is this utility supposed to allow any file to be written to any path (not necessarily the file's correct ufs path)?

gpang commented:
I left some questions about the utility method. Thanks for the awesome feature!

manugoyal commented:
@yupeng9 I think it might be better to leave the tfs as an argument, since the class description is methods on TachyonFileSystem objects. @gpang It needs to be a TachyonFileSystem object, because we use the getInStream method,  which is specific to TachyonFileSystem. I'll put it first though, and also pass in the file info object, rather than any arbitrary path, because it makes more sense to persist to the correct ufsPath.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/6603/
Test PASSed.

gpang commented on clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemUtils.java:
```diff
@@ -134,4 +142,42 @@ public static boolean waitCompleted(final TachyonFileSystemCore tfs, final Tachy
+   * @throws FileDoesNotExistException if the given file does not exist
+   * @throws TachyonException if an unexpected Tachyon error occurs
+   */
+  public static long persistFile(TachyonFileSystem tfs, TachyonFile file, FileInfo fileInfo,
```
I had a question about this utility method. Non-shell clients can call this, right? If so, is it strange that you have to give it a File and a FileInfo for that file?

Also, would the TachyonConf always be the conf from ClientContext?

manugoyal replied:
Yeah I put that in because the shell needed to call ``getFileInfo`` earlier to determine whether we were persisting a directory or a file, so I didn't want to have to make a redundant RPC call.

I guess currently the TachyonConf is whatever the shell passes it, which seems like it's just a default TachyonConf (so not from ``ClientContext``), but since client methods can now call this, it could probably come from a ``ClientContext`` too.

calvinjia replied:
@manugoyal Could we just have the redundant rpc call instead of passing FileInfo? In the future we may merge `FileInfo` into `TachyonFile` which may solve this problem. Also, I think we can remove TachyonConf and just use `ClientContext.getConf` since this will not be used by any server side code.

gpang commented:
I have additional questions about the utility method, persistFile()

jsimsa commented on clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemCore.java:
```diff
@@ -166,6 +167,15 @@ TachyonFile openIfExists(TachyonURI path, OpenOptions options) throws IOExceptio
+   * @param file the TachyonFile to persist
+   * @return true if the persist succeeded, false otherwise
+   */
+  boolean persistFile(TachyonFile file, PersistOptions options)
```
why not just `persist`

manugoyal replied:
It only works on files, not directories.

jsimsa replied:
I understand ... the fact that the function takes TachyonFile as an argument should be sufficient to indicate that this only works for files ... unfortunately, TachyonFile can also be a directory =_=

what happens if you pass a TachyonFile that is a directory? on a related note, why not allow persisting directories? and what about recursive option?

manugoyal replied:
The `persistFile` method on the master will fail if you pass it a directory. I didn't want to deal with directories, because I wanted `TfsShell` to handle exceptions differently than we would normally do. Namely, it tries persisting all the files in the directory, even if some of them fail, and then it combines all the failed files into one exception. I thought that would be weird to implement in a non-TfsShell method, because we only do that in TfsShell.

jsimsa replied:
Sounds good.

jsimsa commented on clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemUtils.java:
```diff
@@ -36,24 +45,24 @@
 
   // prevent instantiation
-  private TachyonFileSystemUtils() {}
+  private TachyonFileSystemUtils() {
```
can we keep this as it was?

jsimsa commented on clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemUtils.java:
```diff
@@ -36,24 +45,24 @@
-   * @throws IOException in case there are problems contacting the Tachyonmaster
-   * for the file status
-   * @throws TachyonException if a Tachyon Exception occurs
+   * @throws IOException          in case there are problems contacting the Tachyonmaster
```
this seems odd

jsimsa commented on clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemCore.java:
```diff
@@ -166,6 +167,15 @@ TachyonFile openIfExists(TachyonURI path, OpenOptions options) throws IOExceptio
       TachyonException;
 
   /**
+   * Persists a {@link TachyonFile} to the UnderFileSystem.
```
`Persist the given file to the under file system.`

jsimsa commented on clients/unshaded/src/main/java/tachyon/client/FileSystemMasterClient.java:
```diff
@@ -468,4 +468,29 @@ public synchronized boolean unmount(TachyonURI tachyonPath) throws IOException {
+    while (!mClosed && (retry ++) <= RPC_MAX_NUM_RETRY) {
+      connect();
+      try {
+        return mClient.persistFile(fileId, length);
```
I have a pending PR that gets rid of the persistFile RPC. We should instead use the "setState" RPC here.

manugoyal replied:
But doesn't `setState` not propagate upwards?

jsimsa replied:
Currently, setState does not support setting the persist bit at all. I am suggesting we do that (along with the persistence propagation).

manugoyal replied:
Oh does your PR do that, or should I do it here?

jsimsa replied:
My PR does not do that. It would be great if you could avoid using the RPC that will soon be removed.

jsimsa commented on clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemUtils.java:
```diff
@@ -134,4 +142,42 @@ public static boolean waitCompleted(final TachyonFileSystemCore tfs, final Tachy
+      TachyonURI dstPath = new TachyonURI(fileInfo.getUfsPath());
+      UnderFileSystem ufs = UnderFileSystem.get(dstPath.getPath(), tachyonConf);
+      String parentPath = dstPath.getParent().getPath();
+      if (!ufs.exists(parentPath) && !ufs.mkdirs(parentPath, true)) {
```
If we use the setState RPC, we could let the server-side to persist the file, which is consistent with other code paths that persist objects created through Tachyon.

jsimsa replied:
Why not push this logic into the setState function on server-side?

manugoyal replied:
Hmm, it seems like there's a lot of client code, with the reading and writing from streams, which I'm not sure really belongs in the `FileSystemMaster` code. Also it seems like other stuff, like Lineage, persists on the worker side. What do you think?

jsimsa replied:
You bring up a good point. This logic probably does not belong into FileSystemMaster. Having said that, we are trying to minimize / avoid UFS operations in the client (for reasons such as avoiding dependencies and improving ease of porting to other languages). Please add a TODO to move this logic to the worker.

calvinjia replied:
@manugoyal could you add the TODO please? The main motivation is to limit client and ufs interactions.

jsimsa commented on clients/unshaded/src/main/java/tachyon/client/file/options/PersistOptions.java:
```diff
@@ -0,0 +1,50 @@
+/*
```
add unit tests for this class

manugoyal commented:
Addressed comments. I reworked the `persistFile` method to use the `setState` RPC call, rather than `persistFile`, so we can remove all the extra stuff I added to the client code, but I also had to modify the `setState` options and related classes to include a `persisted` flag.

jsimsa commented:
please resolve conflicts

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Tachyon-Pull-Request-Builder/6707/
Test FAILed.

jsimsa commented on clients/unshaded/src/test/java/tachyon/client/file/options/SetStateOptionsTest.java:
```diff
@@ -25,13 +25,16 @@
   public void builderTest() {
     Random random = new Random();
     boolean recursive = random.nextBoolean();
+    boolean persist = random.nextBoolean();
```
it looks like the TTL option is not tested ... could you please fix that

jsimsa commented on servers/src/main/java/tachyon/master/file/FileSystemMaster.java:
```diff
@@ -299,8 +301,10 @@ boolean persistFileInternal(long fileId, long length, long opTimeMs)
 
     if (!file.isPersisted()) {
       file.setPersisted(true);
+      propagatePersisted(file, false);
```
is this a bug fix?

manugoyal replied:
Well I copied it over from `persistFile`, which we're not using anymore.

jsimsa replied:
Ack.

jsimsa commented:
thanks @manugoyal, left you some comments

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Tachyon-Pull-Request-Builder/6708/
Test PASSed.

gpang commented on servers/src/main/java/tachyon/master/file/FileSystemMaster.java:
```diff
@@ -1414,9 +1422,22 @@ private void setStateInternal(long fileId, long opTimeMs, SetStateOptions option
+          PreconditionMessage.FILE_TO_PERSIST_MUST_BE_COMPLETE);
+      InodeFile file = (InodeFile) inode;
+      if (!file.isPersisted()) {
+        file.setPersisted(true);
```
This code assumes that the input persisted option flag is always true? Could that be confusing?

gpang commented:
I just had a simple question about the persisted flag in setState(). It looks like no matter what you pass in for 'persisted', it will always persist the file. I think that is confusing, so I think at this point, we can just throw an exception for the "false" case, and add a TODO to figure out what should happen in the un-persist case.

Thanks!

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Tachyon-Pull-Request-Builder/6809/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Tachyon-Pull-Request-Builder/6946/
Test FAILed.

calvinjia commented:
Jenkins, test this please.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Tachyon-Pull-Request-Builder/6948/
Test PASSed.

calvinjia commented on clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemUtils.java:
```diff
@@ -135,4 +143,43 @@ public static boolean waitCompleted(final TachyonFileSystemCore tfs, final Tachy
+      }
+      OutputStream out = closer.register(ufs.create(dstPath.getPath()));
+      ret = IOUtils.copyLarge(in, out);
+    } catch (Throwable e) {
```
Could we catch `Exception` instead here?

calvinjia commented on clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemUtils.java:
```diff
@@ -135,4 +143,43 @@ public static boolean waitCompleted(final TachyonFileSystemCore tfs, final Tachy
+    } finally {
+      closer.close();
+    }
+    // Tell the master to persist the file
```
This is a bit misleading, maybe `mark the file as persisted`?

calvinjia commented on clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemUtils.java:
```diff
@@ -135,4 +143,43 @@ public static boolean waitCompleted(final TachyonFileSystemCore tfs, final Tachy
+   * @param file the file to persist
+   * @param fileInfo the file info of the file
+   * @param tachyonConf TachyonConf object
+   * @return the size of the file persisted
```
Whats the motivation for returning the file size?

manugoyal replied:
It's not used right now, but if this utility method is to be used later, it could be a nice way of seeing how many bytes were copied. @gpang suggested it.

calvinjia commented on servers/src/main/java/tachyon/master/file/FileSystemMaster.java:
```diff
@@ -1372,9 +1376,25 @@ private void setStateInternal(long fileId, long opTimeMs, SetStateOptions option
+        propagatePersisted(file, false);
+      }
+      file.setLastModificationTimeMs(opTimeMs);
+      MasterContext.getMasterSource().incFilesCheckpointed();
```
Should this be in the `!file.isPersisted()` block?

manugoyal replied:
yup

calvinjia commented on servers/src/main/java/tachyon/master/file/FileSystemMaster.java:
```diff
@@ -1372,9 +1376,25 @@ private void setStateInternal(long fileId, long opTimeMs, SetStateOptions option
+        file.setPersisted(true);
+        propagatePersisted(file, false);
+      }
+      file.setLastModificationTimeMs(opTimeMs);
```
Should this be modified in cases where the TTL or pin status was updated?

manugoyal replied:
I would think so. I guess I can put it in all three blocks.

calvinjia commented on shell/src/main/java/tachyon/shell/TfsShell.java:
```diff
@@ -733,7 +769,7 @@ public int getNumOfArgs(String cmd) {
         || cmd.equals("location") || cmd.equals("report") || cmd.equals("pin")
         || cmd.equals("unpin") || cmd.equals("free") || cmd.equals("du") || cmd.equals("unmount")
-        || cmd.equals("loadMetadata") || cmd.equals("unsetTTL")) {
+        || cmd.equals("loadMetadata") || cmd.equals("unsetTTL") || cmd.equals("persist")) {
```
Could you update the docs with the new command?

calvinjia commented:
@manugoyal Left a few comments

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Tachyon-Pull-Request-Builder/6956/
Test PASSed.

calvinjia commented:
@manugoyal Looks good overall, could you add the TODO requested by Jiri?

manugoyal commented:
@calvinjia I think I already added it. That diff is outdated, so it doesn't show up.

calvinjia commented:
LGTM

yupeng9 commented:
LGTM

apc999 commented:
LGTM
